### PR TITLE
Updated to be able to use multi layer in Decoder

### PR DESCRIPTION
**Pull request by kan-bayashi**
Under construction.

Now debugging
- [x] train with dlayers=1
- [x] recog with dlayers=1
- [x] recog_beam  with dlayers=1
- [x] train with dlayers=2
- [x] recog with dlayers=2
- [x] recog_beam  with dlayers=2

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Now it works.
I will check the full performance with voxforge recipe.

File: src/nets/e2e_asr_attctc.py [hunk 1]
```diff
@@ -388,7 +388,7 @@ class Decoder(chainer.Chain):
     def __init__(self, eprojs, odim, dlayers, dunits, sos, eos, att, verbose=0, char_list=None):
         super(Decoder, self).__init__()
         with self.init_scope():
-            self.add_link('emded', L.EmbedID(odim, dunits))
+            self.add_link('embed', L.EmbedID(odim, dunits))
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc.py]: ここは少なくともadd_linkではなく
普通にL.EmbedID(odim, dunits)を定義すればいいと思います
add_linkが必要なのはfor文で回して階層のindexごとのlayerを作りたいからです

File: src/nets/e2e_asr_attctc.py [hunk 2]
```diff
@@ -388,13 +388,15 @@ class Decoder(chainer.Chain):
...
         with self.init_scope():
-            self.embed = L.EmbedID(odim, dunits)
-            # TODO use multiple layers with dlayers option
-            self.decoder = L.StatelessLSTM(dunits + eprojs, dunits)  # 310s per 100 ite -> 240s from NStepLSTM
-            self.output = L.Linear(dunits, odim)
+            self.add_link('embed', L.EmbedID(odim, dunits))
+            self.add_link('lstm0', L.StatelessLSTM(dunits + eprojs, dunits))
+            for l in six.moves.range(1, dlayers):
+                self.add_link('lstm%d' % l, L.StatelessLSTM(dunits, dunits))
+            self.add_link('output', L.Linear(dunits, odim))
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc.py]: 同じコメント
add_linkを使う必要はないです

File: src/nets/e2e_asr_attctc.py (hunk 1 above)
Comment (kan-bayashi) [reply src/nets/e2e_asr_attctc.py]: `add_link`で統一したほうが綺麗かなと思ったのですが
まぜこぜで書いたほうが良いですか？

Comment (sw005320) [reply src/nets/e2e_asr_attctc.py]: なるほどそういう意図ですが
でも個人的にはadd_linkは推奨されていないのでできるだけ少なくした方がいいかなと思います

Comment (kan-bayashi) [reply src/nets/e2e_asr_attctc.py]: 了解です．修正します．

**Comment by kan-bayashi** [comment]
`beam_size = 1` results
```
# for validation set
|       Sum/Avg                     |       1083               79198       |       88.0               6.6                5.4                4.7               16.7               97.8       |
# for evaluation set
|       Sum/Avg                     |       1056               78029       |       88.1               6.5                5.3                4.2               16.1               98.9       |
```

`beam_size = 20` results
```
# for validation set
|       Sum/Avg                     |       1083               79198       |       17.5                2.5               79.9                0.4               82.8              100.0       |
# for evaluation set
|       Sum/Avg                     |       1056               78029       |       16.4                2.2               81.4                0.3               84.0              100.0       |
```
It looks `recog_beam` contains a bug...
However, it is surprised that `beam_size=1` results is better than the score in `RESULTS`.

File: src/nets/e2e_asr_attctc.py [hunk 3]
```diff
@@ -439,10 +444,12 @@ def __call__(self, hs, ys):
 
         # loop for an output sequence
         for i in six.moves.range(olength):
-            att_c, att_w = self.att(hs, z, att_w)
+            att_c, att_w = self.att(hs, z_list[-1], att_w)
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc.py]: it would be better to use z_list[0] (hidden state in a decoder bottom layer)

File: src/nets/e2e_asr_attctc.py [hunk 4]
```diff
@@ -389,12 +389,14 @@ def __init__(self, eprojs, odim, dlayers, dunits, sos, eos, att, verbose=0, char
         super(Decoder, self).__init__()
         with self.init_scope():
             self.embed = L.EmbedID(odim, dunits)
-            # TODO use multiple layers with dlayers option
-            self.decoder = L.StatelessLSTM(dunits + eprojs, dunits)  # 310s per 100 ite -> 240s from NStepLSTM
+            self.add_link('lstm0', L.StatelessLSTM(dunits + eprojs, dunits))
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc.py]: this can be
`self.lstm0 = L.StatelessLSTM(dunits + eprojs, dunits)`

File: src/nets/e2e_asr_attctc.py [hunk 5]
```diff
@@ -439,10 +444,12 @@ def __call__(self, hs, ys):
 
         # loop for an output sequence
         for i in six.moves.range(olength):
-            att_c, att_w = self.att(hs, z, att_w)
+            att_c, att_w = self.att(hs, z_list[-1], att_w)
             ey = F.hstack((eys[:, i, :], att_c))  # utt x (zdim + hdim)
-            c, z = self.decoder(c, z, ey)
-            z_all.append(z)
+            c_list[0], z_list[0] = self['lstm0'](c_list[0], z_list[0], ey)
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc.py]: this can be
`c_list[0], z_list[0] = self.lstm0(c_list[0], z_list[0], ey)`

**Comment by sw005320** [comment]
The latest RESULTS (see below) is still better than 1-best of yours.
```
# change NStepLSTM to StatelessLSTM
$ grep Avg exp/tr_it_a12/decode_*t_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
exp/tr_it_a12/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt:| Sum/Avg               | 1080   78951 | 87.6    6.0    6.4    3.1   15.5   98.5 |
exp/tr_it_a12/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt:| Sum/Avg               | 1050   77586 | 87.3    5.7    7.0    2.9   15.6   98.9 |
```

**Comment by kan-bayashi** [comment]
Sorry my misunderstanding. 
I found the problem when copying z_list and c_list to new_hyp.
```python
# correctly works
                    new_hyp['z_prev'] = []
                    new_hyp['c_prev'] = []
                    for l in six.moves.range(self.dlayers):
                         new_hyp['z_prev'].append(z_list[l])
                         new_hyp['c_prev'].append(c_list[l])
# wrongly works
                    new_hyp['z_prev'] = z_list
                    new_hyp['c_prev'] = c_list
```
Now `recognize_beam` works fine.

**Comment by kan-bayashi** [comment]
`beam_size=20` results
```
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
|       Sum/Avg                     |       1056               78029       |       87.8                5.7                6.5                3.0               15.2               98.5       |
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
|       Sum/Avg                     |       1083               79198       |       88.2                5.8                6.0                3.2               15.0               97.9       |
Finished
```
Became slightly better.

**Comment by sw005320** [comment]
Thanks.
I think this is because the wrong case just gives a pointer, and it would be changed when we use the other hypothesis. We should be careful about keeping the state information in the beam search.
BTW, the result is based on one decoder layer? Did you try more than one decoder layer?

**Comment by kan-bayashi** [comment]
The result is `dlayers=1`.
Now I'm building the model with `dlayers=2`.

**Comment by sw005320** [comment]
Great.
After you did it, and also after I play with it, I'll merge it.
Since it touches a decoder part, it should be careful.

**Comment by kan-bayashi** [comment]
`dlayers=2` and `beam_size=1` results
```
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d2_unit300_location_aconvc10_aconvf100_mtlalpha0.5
_adadelta_bs30_mli800_mlo150/decode_et_it_beam1_eacc.best_p0_len0.0-0.8/result.txt
|       Sum/Avg                     |       1056               78029       |       81.6               9.5                8.9                9.1               27.5               99.3       |
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d2_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_dt_it_beam1_eacc.best_p0_len0.0-0.8/result.txt
|       Sum/Avg                     |       1083               79198       |       81.6               9.9                8.5                9.2               27.6               99.4       |
```

`dlayers=2` and `beam_size=20` results
```
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d2_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
|       Sum/Avg                     |       1056               78029       |       80.7                7.3               12.0                3.1               22.4               98.9       |
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d2_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
|       Sum/Avg                     |       1083               79198       |       80.4                7.5               12.1                3.4               23.1               99.3       |
```

training log
loss
acc

**Comment by sw005320** [comment]
@kan-bayashi can you apply it to pytorch backend?

**Comment by kan-bayashi** [comment]
@sw005320  OK, I will do it.

**Comment by kan-bayashi** [comment]
Finished.

**Comment by sw005320** [comment]
@ShigekiKarita please review this PR.

**Comment by kan-bayashi** [comment]
Now it's time to review.

**Comment by kan-bayashi** [comment]
I have no idea about failure in test_initialization.
If you have any idea, please share.

**Comment by ShigekiKarita** [comment]
OK. I will review this.

@kan-bayashi it sometimes fails randomly because weight-initialization is random. Maybe it is better to add `@pytest.mark.flaky(reruns=5)`.  https://github.com/pytest-dev/pytest-rerunfailures#re-run-individual-failures As a workaround, you can push `rebuild job` button in travis-ci.

BTW, you still have pep8 violations in
https://travis-ci.org/espnet/espnet/jobs/321416534#L476

**Comment by kan-bayashi** [comment]
Voxforge recipe seem to be worked well.
Here is the results.
```
# vggblstmp dlayers = 1 with chainer 
write a result in exp/tr_it_d1_debug_chainer/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1055   77966 | 87.5    5.8    6.6    2.8   15.3   98.9 |
write a result in exp/tr_it_d1_debug_chainer/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1082   79133 | 87.1    5.9    7.0    3.1   16.0   98.9 |

# vggblmstp dlayers = 2 with chainer
write a result in exp/tr_it_d2_debug_chainer/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1082   79133 | 79.4    7.9   12.7    3.3   23.9   99.4 |
write a result in exp/tr_it_d2_debug_chainer/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1055   77966 | 78.4    7.7   13.9    3.2   24.8   98.9 |

# blstmp dlayers = 1 with pytorch
write a result in exp/tr_it_d1_debug_pytorch/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1055   77966 | 84.8    7.1    8.1    3.2   18.4   98.9 |
write a result in exp/tr_it_d1_debug_pytorch/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1082   79133 | 85.2    7.2    7.6    3.3   18.1   98.5 |

# blstmp dlayers = 2 with pytorch
write a result in exp/tr_it_d2_debug_pytorch/decode_dt_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1082   79133 | 76.7    9.0   14.4    4.0   27.4   99.7 |
write a result in exp/tr_it_d2_debug_pytorch/decode_et_it_beam20_eacc.best_p0_len0.0-0.8/result.txt
| SPKR                  | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| Sum/Avg               | 1055   77966 | 75.3    8.7   16.0    3.9   28.6   99.3 |
```

**Comment by sw005320** [comment]
Is it possible to use a same seed for weight-initialization? I thought that I did it, but maybe not. Also, L.embed does not produce the same result even with the same machine due to the atomic operation, and we need some modifications (but as far as I understand, the other parts should be OK).

**Comment by sw005320** [comment]
Yes, this seems to be normal degradation. After we passed the check, we can merge this.

**Comment by kan-bayashi** [comment]
@ShigekiKarita Thank you for your adivice! I will re-check it.
@sw005320 I think in the test scripts, fixing seed is not performed.
I will add the seed in the test script.

**Comment by sw005320** [comment]
I see. Adding the seed makes the test stable, so let's do it.

File: src/nets/e2e_asr_attctc.py [hunk 6]
```diff
@@ -412,13 +412,14 @@ def __init__(self, eprojs, odim, dlayers, dunits, sos, eos, att, verbose=0, char
         super(Decoder, self).__init__()
         with self.init_scope():
             self.embed = L.EmbedID(odim, dunits)
-            # TODO(watanabe) use multiple layers with dlayers option
-            # 310s per 100 ite -> 240s from NStepLSTM
-            self.decoder = L.StatelessLSTM(dunits + eprojs, dunits)
+            self.lstm0 = L.StatelessLSTM(dunits + eprojs, dunits)
+            for l in six.moves.range(1, dlayers):
+                self.add_link('lstm%d' % l, L.StatelessLSTM(dunits, dunits))
```
Comment (ShigekiKarita) [inline src/nets/e2e_asr_attctc.py]: I think it is better to use setattr(self, 'lstm%d' % l, L.StatelessLSTM(dunits, dunits)) .
add_link is deprecated as seen in https://github.com/chainer/chainer/blob/7dbf31f8cea8e7c7229a41bd7ee3dd1896a2e71e/chainer/link.py#L708

Comment (sw005320) [reply src/nets/e2e_asr_attctc.py]: Great suggestions. We have used add_link in the other parts. @kan-bayashi can you also change these parts to `setattr()`?

Comment (kan-bayashi) [reply src/nets/e2e_asr_attctc.py]: OK! I replace `add_link` with yours.

**Comment by ShigekiKarita** [comment]
@kan-bayashi you can use `torch.manual_seed(1)` in torch http://pytorch.org/docs/master/torch.html#torch.manual_seed

**Comment by ShigekiKarita** [comment]
I forgot to add `autopep8  ... --max-line-length 120` option. I think now it is up to 79 chars in travis
https://github.com/kan-bayashi/espnet/blob/a287226f613ca068d6b2a185a9660bc6a07807f4/.travis.yml#L29

**Comment by kan-bayashi** [comment]
Done!!!

**Comment by sw005320** [comment]
LGTM
